File: haskell-server/src/BoardGame/Server/Domain/PieceProvider.hs
--
-- Copyright 2017-2018 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
--

{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE DisambiguateRecordFields #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE ScopedTypeVariables #-}

module BoardGame.Server.Domain.PieceProvider (
    PieceProvider(..)
  , BoardGame.Server.Domain.PieceProvider.take
  , takePieces
  , swapOne
  , pieceProviderType
  , mkDefaultCyclicPieceProvider
  )
  where

import Control.Monad.IO.Class (MonadIO(..))
import Control.Monad.Except (MonadError(..))

import BoardGame.Common.Domain.Piece (Piece, Piece(Piece))
import qualified BoardGame.Common.Domain.PieceProviderType as PieceProviderType
import BoardGame.Common.Domain.PieceProviderType
import BoardGame.Server.Domain.GameError (GameError)

-- The piece generator types are closed in this implementation.
-- TODO. Would be nice to have an open piece generator implementation model.
-- Had some type system issues using type classes.

-- TODO. Name cyclic constructor parameters.

-- | Piece generator.
--   Included in the common package to allow client tests
--   to generate pieces consistently with the server.
data PieceProvider =
  RandomPieceProvider { counter :: Integer, randomizer :: IO Char} |
  CyclicPieceProvider Integer String

take :: (MonadError GameError m, MonadIO m) => PieceProvider -> m (Piece, PieceProvider)

take (provider @ RandomPieceProvider {counter, randomizer}) = do
  letter <- liftIO $ randomizer
  let piece = Piece letter (show counter)
  let nextProvider = RandomPieceProvider (counter + 1) randomizer
  return (piece, nextProvider)

take (CyclicPieceProvider count cycler) = do
  let count' = count + 1
      piece = Piece (head cycler) (show count')
  return (piece, CyclicPieceProvider count' (drop 1 cycler))

-- TODO. Best practices to disambiguate against Prelude [take]?
take' :: (MonadError GameError m, MonadIO m) => PieceProvider -> m (Piece, PieceProvider)
take' = BoardGame.Server.Domain.PieceProvider.take

takePieces :: (MonadError GameError m, MonadIO m) => PieceProvider -> Int -> m ([Piece], PieceProvider)
takePieces provider max = takePiecesAux provider [] max

takePiecesAux :: (MonadError GameError m, MonadIO m) => PieceProvider -> [Piece] -> Int -> m ([Piece], PieceProvider)
takePiecesAux provider list n =
  if n <= 0
    then return (list, provider)
    else do
      (piece, provider1) <- take' provider
      (pieces, provider2) <- takePiecesAux provider1 (piece:list) (n - 1)
      return (pieces, provider2)

swapOne :: (MonadError GameError m, MonadIO m) => PieceProvider -> Piece -> m (Piece, PieceProvider)
swapOne provider piece = do
  (swappedPiece, provider1) <- take' provider
  -- provider2 <- give provider1 piece
  return (swappedPiece, provider1)

pieceProviderType :: PieceProvider -> PieceProviderType
pieceProviderType (RandomPieceProvider _ _) = PieceProviderType.Random
pieceProviderType (CyclicPieceProvider _ _) = PieceProviderType.Cyclic

caps = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
mkDefaultCyclicPieceProvider :: PieceProvider
mkDefaultCyclicPieceProvider = CyclicPieceProvider 0 (cycle caps)

File: scala-server/app/com/bolour/boardgame/scala/server/domain/PieceProvider.scala
/*
 * Copyright 2017-2018 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
 */
package com.bolour.boardgame.scala.server.domain

import scala.util.{Success, Try}
import com.bolour.boardgame.scala.common.domain.Piece

/**
  * Factory for generating new pieces as pieces are consumed in plays.
  */
sealed abstract class PieceProvider {

  def take(): Try[(PieceProvider, Piece)]

  def takePieces(num: Int): Try[(PieceProvider, List[Piece])] = takePiecesAux(List(), num)

  def takePiecesAux(pieces: List[Piece], n: Int): Try[(PieceProvider, List[Piece])] = {
    if (n == 0)
      return Success((this, pieces))
    for {
      (provider1, piece) <- take()
      (provider2, pieces) <- provider1.takePiecesAux(piece +: pieces, n - 1)
    } yield (provider2, pieces)
  }

  def swapOne(piece: Piece): Try[(PieceProvider, Piece)] = {
    for {
      (provider1, resultPiece) <- this.take()
    } yield (provider1, resultPiece)
  }
}

/**
  * Cyclic piece generator - cycles through a given list of letters generating successive pieces.
  *
  * @param valueList The list of letters to cycle through.
  */
case class CyclicPieceProvider(valueList: String) extends PieceProvider {
  if (valueList == null || valueList.isEmpty)
    throw new IllegalArgumentException("empty cyclic generator value list")
  val cycleLength: Int = valueList.length
  var counter: BigInt = 0

  override def take(): Try[(PieceProvider, Piece)] = Try {
    val index: Int = (counter % cycleLength).toInt
    val value = valueList(index)
    val id = counter.toString
    counter = counter + 1
    (this, Piece(value, id))
  }
}

object CyclicPieceProvider {
  def apply() = new CyclicPieceProvider("ABCDEFGHIJKLMNOPQRSTUVWXYZ")
}

/**
  * Random piece generator - generate pieces with random letters.
  */
case class RandomPieceProvider(randomizer: () => Char) extends PieceProvider {

  var counter: BigInt = 0

  def take(): Try[(RandomPieceProvider, Piece)] = {
    val letter = randomizer()
    val id = counter.toString
    val piece = Piece(letter, id)
    counter = counter + 1
    Try((this, piece))
  }
}

